### ISSUE TITLE
Accessing Parameter Groups on a Material Instance

### ISSUE THREAD

**Issue by matthewkapfhammer** [20tab/UnrealEnginePython#103]
Hiya,

I can query the default material properties without issue via:
```python
>>> import unreal_engine as ue
>>> from pprint import pformat
>>> ue.log(pformat(material.properties()))
['ParameterStateId',
 'PhysMaterial',
 'Parent',
 'bHasStaticPermutationResource',
 'bOverrideSubsurfaceProfile',
 'FontParameterValues',
 'ScalarParameterValues',
 'TextureParameterValues',
 'VectorParameterValues',
 'bOverrideBaseProperties',
 'BasePropertyOverrides',
 'ReferencedTextureGuids',
 'SubsurfaceProfile',
 'LightmassSettings',
 'PreviewMesh',
 'ThumbnailInfo',
 'LightingGuid']
```

Any ideas on how might I also query a material instance's parameter groups and access the values?

Cheers,
MK

**Comment by matthewkapfhammer** [comment]
I rechecked the material in the editor and noticed that most of what I was looking for were scalar or texture materials. 

But I still can't figure out how to access the name that's displayed in the Material Instance Editor. This is the best I can get at the moment:
```python
>>> sclvals = material.get_property('ScalarParameterValues')
>>> for v in sclvals:ue.log(v)
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
<unreal_engine.UObject 'ScalarParameterValue' (0x00000260CCFFF100) UClass 'ScriptStruct'>
>>> for v in sclvals:ue.log(v.get_name())
ScalarParameterValue
ScalarParameterValue
ScalarParameterValue
ScalarParameterValue
ScalarParameterValue
ScalarParameterValue
ScalarParameterValue
ScalarParameterValue
ScalarParameterValue
ScalarParameterValue
ScalarParameterValue
ScalarParameterValue
ScalarParameterValue
ScalarParameterValue
ScalarParameterValue
ScalarParameterValue
>>> for v in sclvals:ue.log(v.get_full_name())
ScriptStruct /Script/Engine.ScalarParameterValue
ScriptStruct /Script/Engine.ScalarParameterValue
ScriptStruct /Script/Engine.ScalarParameterValue
ScriptStruct /Script/Engine.ScalarParameterValue
ScriptStruct /Script/Engine.ScalarParameterValue
ScriptStruct /Script/Engine.ScalarParameterValue
ScriptStruct /Script/Engine.ScalarParameterValue
ScriptStruct /Script/Engine.ScalarParameterValue
ScriptStruct /Script/Engine.ScalarParameterValue
ScriptStruct /Script/Engine.ScalarParameterValue
ScriptStruct /Script/Engine.ScalarParameterValue
ScriptStruct /Script/Engine.ScalarParameterValue
ScriptStruct /Script/Engine.ScalarParameterValue
ScriptStruct /Script/Engine.ScalarParameterValue
ScriptStruct /Script/Engine.ScalarParameterValue
ScriptStruct /Script/Engine.ScalarParameterValue
```
Any ideas on how I might access the names displayed in the GUI?

**Comment by rdeioris** [comment]
Hi, i think an ad-hoc method should be added as the name is something available only in the editor. I'll check it now and i will be back to you

**Comment by matthewkapfhammer** [comment]
I had a feeling it was stored elsewhere. Thank you!

**Comment by rdeioris** [comment]
I have discovered how to access them, i am putting down docs here: https://github.com/20tab/UnrealEnginePython/blob/master/docs/Material_API.md

the last section is the one interesting yout, it should be ready in few minutes

**Comment by rdeioris** [comment]
Done: https://github.com/20tab/UnrealEnginePython/blob/master/docs/Material_API.md#listing-getting-and-chaning-available-material-properties

ensure you are using latest code (new binary release will be issued tomorrow)

**Comment by matthewkapfhammer** [comment]
That's pretty rad. I'll test it out on Monday morning or whenever you have a chance to release. Thanks for making time for this request!!

**Comment by matthewkapfhammer** [comment]
Win 10
UE-4.14.3
UnrealEnginePython_20170115_python27.zip (Note: I couldn't get either of the 3.x versions to work in an existing project. Beyond the scope of this issue, though.)

It's  working, but you have to already know the type of the data in order to know how to access it. Do you know of an existing way to do a type check in order to dynamically call the right getter?

Example:
```python
    def get_map_path(material_instance, parameter_name):
        uobject = material_instance.get_material_texture_parameter(parameter_name)
        return uobject.get_path_name(parameter_name)

    def get_scalar(material_instance, parameter_name):
        return material_instance.get_material_scalar_parameter(parameter_name)

    typed_getter = {
        'Color Map UND': get_map_path,
        'Color Root UND': get_scalar,
        # etc.
    }
    material_parent = material_instance.Parent
    for expression in material_parent.Expressions:
        try:
            parameter_group = expression.Group
            parameter_name = expression.ParameterName
            try:
                parameter_value = typed_getter[parameter_name](material_instance,
                                                               parameter_name)
            except:
                msg = traceback.format_exc()
                ue.log_warning(msg)
                parameter_value = None

            msg = '[{0}]{1}: {2}'.format(
                parameter_group,
                parameter_name,
                parameter_value)
            ue.log(msg)
        except AttributeError:
            msg = traceback.format_exc()
            ue.log_warning(msg)
```

Thanks again for taking on this issue so quickly!

**Comment by rdeioris** [comment]
Hi, you can simply check the class of the expression (you can use get_class()):

```py
if expression.get_class().get_name() == 'MaterialExpressionScalarParameter':
    ...
```

or (more elegant)

```py
from unreal_engine.classes import MaterialExpressionScalarParameter

if expression.get_class() == MaterialExpressionScalarParameter:
    ...
```

**Comment by matthewkapfhammer** [comment]
Ha. That probably should have been more obvious to me. Cheers for that!

I tried to get a full list of param classes via dir, but that doesn't work: 
```python
>>> unreal_engine.log(pformat(dir(unreal_engine.classes)))
['__call__',
 '__class__',
 '__cmp__',
 '__delattr__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__le__',
 '__lt__',
 '__module__',
 '__name__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__self__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__']
```

To get what I needed, I just logged the class name for the expressions I actually had to deal with in my one-off script:
```python
for expression in material_parent.Expressions:
    ue.log(expression.get_class().get_name())
```
Read by get_material_scalar_parameter:
MaterialExpressionScalarParameter
MaterialExpressionStaticBoolParameter
MaterialExpressionStaticSwitchParameter

Read by get_material_texture_parameter:
MaterialExpressionTextureObjectParameter
MaterialExpressionTextureSampleParameter2D

Read by get_material_vector_parameter:
MaterialExpressionVectorParameter


I was able to see the full class list here: 
https://github.com/EpicGames/UnrealEngine/tree/55c9f3ba0010e2e483d49a4cd378f36a46601fad/Engine/Source/Runtime/Engine/Classes/Materials

Anyway, it works. I think we can close this now.
